sys/net/gnrc/network_layer/ndp/gnrc_ndp.c

```diff
@@ -248,11 +218,53 @@ void gnrc_ndp_nbr_adv_handle(kernel_pid_t iface, gnrc_pktsnip_t *pkt,
                 nc_entry->flags &= ~GNRC_IPV6_NC_IS_ROUTER;
                 /* TODO: update FIB */
             }
+
+            while ((queued_pkt = gnrc_pktqueue_remove_head(&nc_entry->pkts)) != NULL) {
+                gnrc_netapi_send(gnrc_ipv6_pid, queued_pkt->pkt);
+                queued_pkt->pkt = NULL;
+            }
         }
-        else if (l2tgt_changed) {
-            if (gnrc_ipv6_nc_get_state(nc_entry) == GNRC_IPV6_NC_STATE_REACHABLE) {
-                gnrc_ndp_internal_set_state(nc_entry, GNRC_IPV6_NC_STATE_STALE);
+        else {
+            /* first or-term: no link-layer, but nc_entry has l2addr,
+             * second or-term: different l2addr cached */
+            bool l2tgt_changed = false;
+
+            if ((!_pkt_has_l2addr(netif_hdr)) && (l2tgt_len == 0)) {
+                /* there was previously a L2 address registered */
+                l2tgt_changed = (nc_entry->l2_addr_len != 0);
+            }
+            /* link-layer has addresses and TLLAO with different address */
+            else if (_pkt_has_l2addr(netif_hdr) && (l2tgt_len != 0)) {
+                l2tgt_changed = (!(l2tgt_len == nc_entry->l2_addr_len)) &&
+                                (memcmp(nc_entry->l2_addr, l2tgt, l2tgt_len) == 0);
+            }
 
+            if ((nbr_adv->flags & NDP_NBR_ADV_FLAGS_O) || !l2tgt_changed ||
+                (l2tgt_len == 0)) {
+                if (l2tgt_len != 0) {
+                    nc_entry->iface = iface;
+                    nc_entry->l2_addr_len = l2tgt_len;
+                    memcpy(nc_entry->l2_addr, l2tgt, l2tgt_len);
+                }
+
+                if (nbr_adv->flags & NDP_NBR_ADV_FLAGS_S) {
+                    gnrc_ndp_internal_set_state(nc_entry, GNRC_IPV6_NC_STATE_REACHABLE);
+                }
+                else if (l2tgt_changed) {
+                    gnrc_ndp_internal_set_state(nc_entry, GNRC_IPV6_NC_STATE_STALE);
+                }
+
+                if (nbr_adv->flags & NDP_NBR_ADV_FLAGS_R) {
+                    nc_entry->flags |= GNRC_IPV6_NC_IS_ROUTER;
+                }
+                else {
+                    nc_entry->flags &= ~GNRC_IPV6_NC_IS_ROUTER;
+                    /* TODO: update FIB */
+                }
+            }
+            else if (l2tgt_changed &&
+                     gnrc_ipv6_nc_get_state(nc_entry) == GNRC_IPV6_NC_STATE_REACHABLE) {
+                gnrc_ndp_internal_set_state(nc_entry, GNRC_IPV6_NC_STATE_STALE);
             }
         }
     }
```

Comment:
this function could also benefit from being split up, but that's a different topic.